private/pkg/storage/storagemanifest/storagemanifest.go

```diff
@@ -0,0 +1,50 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storagemanifest
+
+import (
+	"github.com/bufbuild/buf/private/pkg/manifest"
+	"github.com/bufbuild/buf/private/pkg/storage"
+)
+
+// NewReadBucket takes a Manifest and BlobSet and builds a ReadBucket
+// that contains the files in the Manifest.
+func NewReadBucket(
+	m *manifest.Manifest,
+	blobSet *manifest.BlobSet,
```

Comment:
These two being structs and not interfaces, I'm really uneasy to receive them as pointers because it forces the func to do nil-checking, instead of the caller making sure it's a value and not a pointer. Also, passing the struct gives peace of mind to both caller and function that we're passing a copy, and no changes to the original data will be made.

I know about the rule of function pointer receivers, but from what I understand, that doesn't apply to parameters.

Reply:
We do not do this style in Buf's codebases - this was resolved a while ago. Use pointers unless there is a strict reason not to. In this case, you're saying that by derefencing, you are passing a copy of the `Manifest`, and no changes to the underlying data can be made. This isn't actually what you are doing here - you're passing a reference the the underlying map by value, meaning that i.e. an `AddEntry` call on `manifest.Manifest` will still result in a mutation to the dereferenced `*manifest.Manifest` you passed to the function.

As a demonstration of this issue, run the following

```go
package main

import (
	"fmt"
	"sort"
)

func main() {
	foo := NewFoo()
	AddToDereferencedFoo(*foo, "one")
	AddToDereferencedFoo(*foo, "two")
	fmt.Println(foo.Bars())
}

func AddToDereferencedFoo(foo Foo, bar string) {
	foo.AddBar(bar)
}

type Foo struct {
	bars map[string]struct{}
}

func NewFoo() *Foo {
	return &Foo{
		bars: make(map[string]struct{}),
	}
}

func (f *Foo) AddBar(bar string) {
	f.bars[bar] = struct{}{}
}

func (f *Foo) Bars() []string {
	bars := make([]string, len(f.bars))
	for bar := range f.bars {
		bars = append(bars, bar)
	}
	sort.Strings(bars)
	return bars
}
```

You'll see that it prints out both "one" and "two".

It's this reason (among many others) that as an organization, we use pointers unless there is a strict reason not to - subtle bugs get introduced where people are relying on pass-by-value semantics an but when there's actually pass-by-reference semantics under the hood, and developers do not fully consider the semantics, and get a sense of safety when they shouldn't.

If you want immutability to be a property, have an interface that doesn't expose the mutable methods.

Reply:
I'd probably refactor `Manifest` to be an interface, but I am not advocating that we need to do that. But yes of note.